### DISCUSSION TITLE
Anyway to get a Union type to work in a operation?

### DISCUSSION THREAD

**Discussion by Devon-White** [microsoft/typespec#4460] [OpenAPI3] [answered]
I'm trying to accomplish the ability to pass a Union type in a operation.

It seems to only take `Model` types. So, i'm having a hard time declaring a operation that can take one of two types.

Essentially, in the API, there are two params that cannot be passed together, but work if passed individually.

Due to a operation only allowing me to pass in a model, i'm only able to pass in one of these Types.

I can create a type, and pass a property that takes in the union, but then I have a property that does not exist.

Example can be found [here](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaTMiOw0K0h5odHRwxBoNCnVzaW5nIFR5cGVTcGVjLk9wZW5BUEk70RlIdHRwxTFAc2VydmljZSh7DQogIHRpdGxlOiAiVGVzdCBBUEkiLA0KfSnHJmVyKCLEanM6Ly90ZXN0L2FwaS8iLCAiRW5kcG9pbnQiLCB7xS51c2VBdXRoKEJhc2ljxArEFXJvdXRlKCLFOyIpDQpuYW1lc3BhY2UgxG5BUEkg5QCDbW9kZWzFE0Jhc2XGFCAgYTogImHkAOPEDWI6ICJiyA1jOiAiY8YNfeQA0cxCQSBleHRlbmRz0VBkOiAiZMhD5ADyZddDQtlDZjogImbIQ2c6ICJn10NDyDLkAPpfdW5pb246xRlVxAvMNC8qzjpEyDouLi7EEEEgfMgLQsc3ICAqL8Y9QG9uZU9mxArFY8piyEXFQizKDELFDMd7YWxpYeYA5EHFCj3GDCB8xivlAiwgIGludGVyZuQBtFNJUOgB93PIYWxpc3Qo5wCgQyk6IHZvaWToAN99DQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

I'm hoping to accomplish something similar to the spec below. I want the ability to declare a request body of a operation as a Union that can declare to use `oneOf` the listed `refs`.

```yaml
openapi: 3.0.0
info:
  title: Test API
  version: 0.0.0
tags: []
paths:
  /test:
    post:
      operationId: SIPEndpoints_list
      parameters: []
      responses:
        '204':
          description: 'There is no content to send for this request, but the headers may be useful. '
      requestBody:
        required: true
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/TestUnion'
security:
  - BasicAuth: []
components:
  schemas:
    TestA:
      type: object
      required:
        - a
        - b
      properties:
        a:
          type: string
          enum:
            - a
        b:
          type: string
          enum:
            - b
    TestB:
      type: object
      required:
        - c
        - d
      properties:
        c:
          type: string
          enum:
            - c
        d:
          type: string
          enum:
            - d
    TestUnion:
      oneOf:
        - $ref: '#/components/schemas/TestA'
        - $ref: '#/components/schemas/TestB'
  securitySchemes:
    BasicAuth:
      type: http
      scheme: basic
servers:
  - url: https://test/api/
    description: Endpoint
    variables: {}
```

SWAGGER editor pic: 
![image](https://github.com/user-attachments/assets/df3f21a6-6d01-4f87-a8a1-d449c0f64518)

**Comment by mario-guerra** [comment]
Hi @Devon-White, [this looks pretty close](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaTMiOw0K0h5odHRwxBoNCnVzaW5nIFR5cGVTcGVjLk9wZW5BUEk70RlIdHRwxTFAc2VydmljZSh7DQogIHRpdGxlOiAiVGVzdCBBUEkiLA0KfSnHJmVyKCLEanM6Ly90ZXN0L2FwaS8iLCAiRW5kcG9pbnQiLCB7xS51c2VBdXRoKEJhc2ljxArEFXJvdXRlKCLFOyIpxHxuYW1lc3BhY2UgxHBBUEkg5QCFbW9kZWzFE0Jhc2XGFCAgYTogImHkAOXEDWI6ICJiyA1jOiAiY8YNfcRXzEJBIGV4dGVuZHPRUGQ6ICJkyEPkAPRl10NC2UNmOiAiZshDZzogImfXQ0PIMuQA%2FF91bmlvbjrFGVXEC8w0LyrOOkTIOi4uLsQQQSB8yAtCxzcgICovxj1Ab25lT2bECsVjymLIRcVCLMoMQsUMx3thbGlh5gDkQcUKPcYMIHzGK%2BUCLsVmZ2V0xAhvcCBTSVDoAfxzX2xpc3QoKTrIbEBzdGF0dXNDb2RlIMoLOiAyMDTnATJAYm9keSBjb250ZW508QEQxXB9DQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D), does it give you what you need?

**Reply by Devon-White** [reply]
> Hi @Devon-White, [this looks pretty close](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaTMiOw0K0h5odHRwxBoNCnVzaW5nIFR5cGVTcGVjLk9wZW5BUEk70RlIdHRwxTFAc2VydmljZSh7DQogIHRpdGxlOiAiVGVzdCBBUEkiLA0KfSnHJmVyKCLEanM6Ly90ZXN0L2FwaS8iLCAiRW5kcG9pbnQiLCB7xS51c2VBdXRoKEJhc2ljxArEFXJvdXRlKCLFOyIpxHxuYW1lc3BhY2UgxHBBUEkg5QCFbW9kZWzFE0Jhc2XGFCAgYTogImHkAOXEDWI6ICJiyA1jOiAiY8YNfcRXzEJBIGV4dGVuZHPRUGQ6ICJkyEPkAPRl10NC2UNmOiAiZshDZzogImfXQ0PIMuQA%2FF91bmlvbjrFGVXEC8w0LyrOOkTIOi4uLsQQQSB8yAtCxzcgICovxj1Ab25lT2bECsVjymLIRcVCLMoMQsUMx3thbGlh5gDkQcUKPcYMIHzGK%2BUCLsVmZ2V0xAhvcCBTSVDoAfxzX2xpc3QoKTrIbEBzdGF0dXNDb2RlIMoLOiAyMDTnATJAYm9keSBjb250ZW508QEQxXB9DQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D), does it give you what you need?

I dont believe so, this seems to be for the `response`, but I need this for the `request body`.

This generates with no request body:

![image](https://github.com/user-attachments/assets/df6aea16-5ee8-48e7-93c4-a6a726b009c6)

For my use case, the response will not be a union, and will return both properties, the request is the one that needs to be a union.

**Comment by timotheeguerin** [comment, marked as answer]
You just want to add `@body` on the property to say the body is the content of that prop, then it can be anything 
[Playground](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaTMiOw0K0h5odHRwxBoNCnVzaW5nIFR5cGVTcGVjLk9wZW5BUEk70RlIdHRwxTFAc2VydmljZSh7DQogIHRpdGxlOiAiVGVzdCBBUEkiLA0KfSnHJmVyKCLEanM6Ly90ZXN0L2FwaS8iLCAiRW5kcG9pbnQiLCB7xS51c2VBdXRoKEJhc2ljxArEFXJvdXRlKCLFOyIpDQpuYW1lc3BhY2UgxG5BUEkg5QCDbW9kZWzFE0Jhc2XGFCAgYTogImHkAOPEDWI6ICJiyA1jOiAiY8YNfeQA0cxCQSBleHRlbmRz0VBkOiAiZMhD5ADyZddDQtlDZjogImbIQ2c6ICJnzUNAb25lT2bECnVuaW9uxT9VxQrHQMQRQSzKDELFDMc%2BYWxpYcZzQcUKPcYMIHzGK%2BUBuyAgaW50ZXJm5AFDU0lQ6AGGc8hhbGlzdChAYm9keSDkAXRf5QCIOuoAiSk6IHZvaWToALV9DQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

**Reply by Devon-White** [reply]
> You just want to add `@body` on the property to say the body is the content of that prop, then it can be anything [Playground](https://typespec.io/playground?c=aW1wb3J0ICJAdHlwZXNwZWMvb3BlbmFwaTMiOw0K0h5odHRwxBoNCnVzaW5nIFR5cGVTcGVjLk9wZW5BUEk70RlIdHRwxTFAc2VydmljZSh7DQogIHRpdGxlOiAiVGVzdCBBUEkiLA0KfSnHJmVyKCLEanM6Ly90ZXN0L2FwaS8iLCAiRW5kcG9pbnQiLCB7xS51c2VBdXRoKEJhc2ljxArEFXJvdXRlKCLFOyIpDQpuYW1lc3BhY2UgxG5BUEkg5QCDbW9kZWzFE0Jhc2XGFCAgYTogImHkAOPEDWI6ICJiyA1jOiAiY8YNfeQA0cxCQSBleHRlbmRz0VBkOiAiZMhD5ADyZddDQtlDZjogImbIQ2c6ICJnzUNAb25lT2bECnVuaW9uxT9VxQrHQMQRQSzKDELFDMc%2BYWxpYcZzQcUKPcYMIHzGK%2BUBuyAgaW50ZXJm5AFDU0lQ6AGGc8hhbGlzdChAYm9keSDkAXRf5QCIOuoAiSk6IHZvaWToALV9DQo%3D&e=%40typespec%2Fopenapi3&options=%7B%7D)

Thank you, works exactly the way I need it too. :)
